# 2.5 Let’s Code!
Sebelum mengakhiri bab ini, mari kita buat kode program dari dua eksperimen yang telah dijelaskan, yaitu identifikasi gambar fashion dan penambahan jumlah layer pada jaringan.

## 2.5.1   Identifikasi Gambar Fashion

Seperti pada eksperimen sebelumnya, pertama import library TensorFlow dan Keras yang dibutuhkan:

In [1]:
from tensorflow.keras import datasets
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten, Dropout,Dense

In [2]:
!nvidia-smi

Thu Sep  4 14:12:55 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 561.19                 Driver Version: 561.19         CUDA Version: 12.6     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                  Driver-Model | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 3060 ...  WDDM  |   00000000:01:00.0 Off |                  N/A |
| N/A   48C    P5             13W /   92W |     117MiB /   6144MiB |     41%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

Unduh dan siapkan dataset yang digunakan untuk pelatihan model. Dalam hal ini sama dengan dataset pada eksperimen sebelumnya:

In [3]:
fashion = datasets.fashion_mnist

(X_train, y_train), (X_test, y_test) = fashion.load_data() 

In [4]:
X_train = X_train/255.0
X_test  = X_test/255.0

#### ANN dengan 1 Hidden Layer dan SGD Optimizer

Siapkan model ANN dengan 1 hidden layer yang memiliki 128 node:


In [5]:
from tensorflow import keras

model1 = keras.models.Sequential()
model1.add(Flatten(input_shape=[28, 28]))
model1.add(Dense(128, activation='relu'))
model1.add(Dense(10, activation='softmax'))


c:\Users\rafif\anaconda3\lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Lakukan kompilasi model menggunakan SGD sebagai optimizer:

In [6]:
model1.compile(loss='sparse_categorical_crossentropy', 
optimizer='SGD',
metrics=['accuracy'])

In [7]:
model1.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 101,770 (397.54 KB)

 Trainable params: 101,770 (397.54 KB)

 Non-trainable params: 0 (0.00 B)

Terdapat total 101,770 paramater yang akan dilatih yang sebagian besar berada pada hidden layer.

Jalankan eksperimen sampai 100 epochs dengan 20% dari Training Set dijadikan sebagai Validation Set.

In [8]:
model1.fit(X_train, y_train, epochs=100, 
validation_split=0.2, 
batch_size=128)

Epoch 1/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.5336 - loss: 1.5175 - val_accuracy: 0.7448 - val_loss: 0.7927
Epoch 2/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7549 - loss: 0.7593 - val_accuracy: 0.7843 - val_loss: 0.6615
Epoch 3/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7887 - loss: 0.6550 - val_accuracy: 0.8011 - val_loss: 0.5986
Epoch 4/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8030 - loss: 0.6042 - val_accuracy: 0.8105 - val_loss: 0.5634
Epoch 5/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8167 - loss: 0.5563 - val_accuracy: 0.8198 - val_loss: 0.5396
Epoch 6/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8220 - loss: 0.5376 - val_accuracy: 0.8256 - val_loss: 0.5198
Epoch 7/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8307 - loss: 0.5106 - val_accuracy: 0.8298 - val_loss: 0.5049
Epoch 8/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8301 - loss: 0.5040 - val_accu

In [9]:
model1.evaluate(X_test, y_test) 

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 924us/step - accuracy: 0.8724 - loss: 0.3669


[0.3708183467388153, 0.868399977684021]

### DNN dengan 2 Hidden Layer dan SGD Optimizer

Siapkan model DNN dengan 2 hidden layer yang masing-masing layer terdiri dari 128 node:

In [10]:
model2 = keras.models.Sequential()
model2.add(Flatten(input_shape=[28, 28]))
model2.add(Dense(128, activation='relu'))
model2.add(Dense(128, activation='relu'))
model2.add(Dense(10, activation='softmax'))

Lakukan kompilasi model dengan menggunakan SGD sebagai optimizer:

In [11]:
model2.compile(loss='sparse_categorical_crossentropy', 
optimizer='SGD', metrics=['accuracy'])

In [12]:
model2.summary() 

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_1 (Flatten)             │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 128)            │        16,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 118,282 (462.04 KB)

 Trainable params: 118,282 (462.04 KB)

 Non-trainable params: 0 (0.00 B)

Terdapat total 118,282 paramater yang akan dilatih dan sebagian besar berada pada hidden layer. Jumlah paramater ini sedikit lebih besar dibanding dengan model1.

Jalankan eksperimen sampai 100 epochs dengan 20% dari Training Set dijadikan sebagai Validation Set:

In [13]:
model2.fit(X_train, y_train, epochs=100,

validation_split=0.2,

batch_size=128)

Epoch 1/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.5206 - loss: 1.5975 - val_accuracy: 0.7157 - val_loss: 0.8072
Epoch 2/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7406 - loss: 0.7659 - val_accuracy: 0.7849 - val_loss: 0.6492
Epoch 3/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7892 - loss: 0.6293 - val_accuracy: 0.8027 - val_loss: 0.5823
Epoch 4/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8063 - loss: 0.5732 - val_accuracy: 0.8154 - val_loss: 0.5394
Epoch 5/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8203 - loss: 0.5254 - val_accuracy: 0.8216 - val_loss: 0.5150
Epoch 6/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8234 - loss: 0.5113 - val_accuracy: 0.8292 - val_loss: 0.4936
Epoch 7/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8344 - loss: 0.4805 - val_accuracy: 0.8296 - val_loss: 0.4843
Epoch 8/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8372 - loss: 0.4694 - val_accu

In [14]:
model2.evaluate(X_test, y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 853us/step - accuracy: 0.8749 - loss: 0.3518


[0.3524436950683594, 0.8747000098228455]

### DNN dengan 2 Hidden Layer dengan Adam Optimizer

Siapkan model DNN dengan 2 hidden layer yang masing-masing layer terdiri dari 128 node. Arsitektur model ini persis sama dengan arsitektur model2:

In [15]:
model3 = keras.models.Sequential()
model3.add(Flatten(input_shape=[28, 28]))
model3.add(Dense(128, activation='relu'))
model3.add(Dense(128, activation='relu'))
model3.add(Dense(10, activation='softmax'))

In [16]:
model3.compile(loss='sparse_categorical_crossentropy', 
optimizer='Adam',
metrics=['accuracy'])

In [17]:
model3.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_2 (Flatten)             │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_6 (Dense)                 │ (None, 128)            │        16,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 118,282 (462.04 KB)

 Trainable params: 118,282 (462.04 KB)

 Non-trainable params: 0 (0.00 B)

In [18]:
model3.fit(X_train, y_train, epochs=100, 
validation_split=0.2, 
batch_size=128)

Epoch 1/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7316 - loss: 0.7776 - val_accuracy: 0.8478 - val_loss: 0.4274
Epoch 2/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8578 - loss: 0.4014 - val_accuracy: 0.8632 - val_loss: 0.3865
Epoch 3/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8716 - loss: 0.3574 - val_accuracy: 0.8758 - val_loss: 0.3541
Epoch 4/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8815 - loss: 0.3277 - val_accuracy: 0.8758 - val_loss: 0.3453
Epoch 5/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8912 - loss: 0.2981 - val_accuracy: 0.8824 - val_loss: 0.3294
Epoch 6/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8962 - loss: 0.2842 - val_accuracy: 0.8809 - val_loss: 0.3321
Epoch 7/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8970 - loss: 0.2742 - val_accuracy: 0.8772 - val_loss: 0.3516
Epoch 8/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9002 - loss: 0.2684 - val_accu

In [19]:
model3.evaluate(X_test, y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 875us/step - accuracy: 0.8826 - loss: 0.8961


[0.8995811939239502, 0.8858000040054321]

### DNN dengan 2 Hidden Layer dengan Adam Optimizer dan Dropout

Siapkan model DNN dengan 2 hidden layer yang masing-masing layer terdiri dari 128 node, serta tambahkan Dropout layer masing-masing setelah input dan hidden layer:

In [20]:
model4 = keras.models.Sequential()
model4.add(Flatten(input_shape=[28, 28]))
model4.add(Dropout(0.3))
model4.add(Dense(128, activation='relu'))
model4.add(Dropout(0.3))
model4.add(Dense(128, activation='relu'))
model4.add(Dropout(0.3))
model4.add(Dense(10, activation='softmax'))

Perhatikan, kita tidak menambahkan Dropout setelah output layer. Lakukan kompilasi model menggunakan Adam sebagai optimizer:

In [21]:
model4.compile(loss='sparse_categorical_crossentropy', 
optimizer='Adam',
metrics=['accuracy'])
 

# Periksa struktur dan parameter model:

model4.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_3 (Flatten)             │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 128)            │        16,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_10 (Dense)                │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 118,282 (462.04 KB)

 Trainable params: 118,282 (462.04 KB)

 Non-trainable params: 0 (0.00 B)

Jumlah paramater tidak berubah, alias sama dengan model3 meskipun telah ditambahkan 3 Dropout layer.

Jalankan eksperimen sampai 100 epochs dengan 20% dari Training Set dijadikan sebagai Validation Set:

In [22]:
model4.fit(X_train, y_train, epochs=100, 
validation_split=0.2, 
batch_size=128)

Epoch 1/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.6042 - loss: 1.0868 - val_accuracy: 0.8096 - val_loss: 0.4973
Epoch 2/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7923 - loss: 0.5684 - val_accuracy: 0.8399 - val_loss: 0.4321
Epoch 3/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8182 - loss: 0.5024 - val_accuracy: 0.8518 - val_loss: 0.4000
Epoch 4/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8294 - loss: 0.4672 - val_accuracy: 0.8590 - val_loss: 0.3844
Epoch 5/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8327 - loss: 0.4538 - val_accuracy: 0.8668 - val_loss: 0.3716
Epoch 6/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8401 - loss: 0.4379 - val_accuracy: 0.8662 - val_loss: 0.3684
Epoch 7/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8459 - loss: 0.4246 - val_accuracy: 0.8690 - val_loss: 0.3617
Epoch 8/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8492 - loss: 0.4067 - val_accu

In [23]:
model4.evaluate(X_test, y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 859us/step - accuracy: 0.8886 - loss: 0.3137


[0.31709524989128113, 0.8863999843597412]